Firestore/Source/API/FIRQuery.m

```diff
@@ -83,6 +83,8 @@ - (instancetype)includeDocumentMetadataChanges:(BOOL)includeDocumentMetadataChan
 
 @interface FIRQuery ()
 @property(nonatomic, strong, readonly) FSTQuery *query;
+- (FIRQuery *)queryFilteredUsingComparisonPredicate:(NSPredicate *)predicate;
```

Comment:
These don't need to be forward declared here.

Anything within the `@implementation` can call anything else without a forward declaration.

To summarize:
  * public properties and methods:
    * declare them in the `@interface` in the main `.h` file
    * define them in the `@implementation` in the `.m`.
  * public methods that should be hidden from a subset of callers:
    * declare them in the `@interface (Internal)` in the main `+Internal.h` file
    * define them in the `@implementation (Internal)` in the `.m`.
    * note this does not work with initializers (they cannot be in any nonempty category).
  * private methods:
    * define them in the `@implementation` in the `.m`
  * private properties:
    * declare them in the empty category (e.g. `@interface FIRQuery ()`) in the `.m`.
    * We have to do this `@property` declarations must be in an `@interface`.
  * protected methods:
    * declare them in the empty category (e.g. `@interface FIRQuery ()`) in the `.m`.
    * define them in the `@implementation` in the `.m`
    * only works if all subclasses are implemented in the same file.

Note that Objective-C really has no concept of access control; this is really just a set of conventions that hide these methods to callers that shouldn't see them.

The reason for the warning originally was that you added these method declarations to `FIRQuery (Internal)` but did not define them in the Internal category. The error was declaring them at all--they only need a definition to function as a private method.

Reply:
Got that! Thank you so much for the detailed instruction!